### PR TITLE
Webhook triggers for configrepo branch support

### PR DESCRIPTION
**Pull request by marques-work**
This PR introduces 2 shiny, new things:

1. A new webhook endpoint, specifically for config repos: `/go/api/webhooks/:provider/config_repos/:id`
2. A UI component that builds the webhook URL for a given config repo to help the user set them up properly on their chosen provider (GitHub, Bitbucket, and friends)

### So how does this help PR/branch support?

Currently, PR support queries your git provider's APIs to refresh the list of pull requests. As such, the resulting pipeline configurations can change even if there are no code changes to your config repo definitions. Because of this, config repos need to poll the APIs regularly, which is probably fine for a small amount of config repos. However, if one has many config repos that take advantage of PR support, this won't scale. What we need are webhooks that notify config repos when a pull request is opened or closed/merged, in addition to when there is a code push so that GoCD *only* queries the API when a real change has happened.

### Context: Why can't the existing webhooks suffice?

Great question.

* Firstly, the current "notify" endpoints do not support any pull/merge request-related events, but only push (and ping, when applicable)
* Secondly, the notify endpoints locate materials by heuristics based on the repository identity reported by the `POST` payload (by guessing the material URL); this won't work because:
    * Config repo materials are not necessarily related to the materials owned by the pipelines one defines in the config repo
    * For example, one might have a disparate private (or filesystem-local) repo that only holds pipeline definitions for that use the GitHub repo you are interested in. The webhook must be able to resolve and trigger the config repo if a push or pull request happens. The config repo need not even be a git repo in this case.
    * Thus, the most straightforward solution is to trigger a specific config repo by its explicit ID, rather than try to figure this out by analyzing the payload

### Notable changes

* Major refactor of webhook domain objects
    * WebhookRequest objects are now decoupled/agnostic to the payload and supported events
        * These are responsible for extracting authentication credentials and event names
        * WebhookRequest objects no longer need to decide payload/event type before inspecting the raw request; previously this made it difficult to support multiple event and payload types in a single endpoint because it's rather catch-22
    * Payload instances tell WebhookRequest instances how to deserialize request bodies; this happens after WebhookRequest instantiation and after event-type identification
    * Validation and authentication now happens in the controller, instead of within the WebhookRequest domain object
        * Much of this logic is in reusable mixins and provider-specific helper classes
* Server configuration site URLs are now exposed to SPA layouts
* New helpers to generate base URLs from server configurations falling back to guessing from `window.location` in the absence of configuration
* New UI component to build webhook URLs for a specific config repo with instructions on what to configure at the `git` provider

### CODE REVIEW

**Comment by marques-work** [comment]
Sorry that this PR is very large and probably hard to review. Most of its largeness is from the webhook domain object refactoring, where classes were split into two or more, renamed, and moved to other packages.

**Review by kritika-singh3 — COMMENTED** [review]
Mostly looks ok to me. Just did a basic code review. 
Testing is pending.

File: api/api-webhook-v1/src/main/java/com/thoughtworks/go/apiv1/webhook/request/payload/pr/GitLabPR.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        return project.hostname();
+    }
+
+    @Override
+    public String fullName() {
+        return project.fullName();
+    }
+
+    private static class Attrs {
+        @SerializedName("iid")
```
[the remark this answers is not in the document]
Comment (kritika-singh3) [reply api/api-webhook-v1/src/main/java/com/thoughtworks/go/apiv1/webhook/request/payload/pr/GitLabPR.java:54]: Sorry ignore me - I was thinking this was a type - but we do have both `id` and `iid`.

**Comment by kritika-singh3** [comment]
Tested via YAML plugin and
 - github
 - gitlab
 - bitbucket cloud.

Works as expected.

File: server/src/main/webapp/WEB-INF/rails/webpack/views/components/buttons/index.tsx [hunk 2]
```diff
@@ -60,9 +64,9 @@ function dataTestIdAttrs(attrs: Attrs) {
 }
 
 abstract class Button extends RestyleViewComponent<Styles, Attrs> {
-  css: Styles                                  = defaultStyles;
+  css: Styles = defaultStyles;
   ajaxOperationMonitor: Stream<OperationState> = Stream<OperationState>(OperationState.UNKNOWN);
-  forceSpinner                                 = false;
+  forceSpinner = false;
```
[the remark this answers is not in the document]
Comment (marques-work) [reply server/src/main/webapp/WEB-INF/rails/webpack/views/components/buttons/index.tsx:69]: Sorry, that was vscode autoformatting with the typescript plugin. I can fix this.

File: server/src/main/webapp/WEB-INF/rails/webpack/views/pages/config_repos/config_repos_widget.tsx [hunk 3]
```diff
@@ -14,38 +14,41 @@
...
+import { CRResult } from "./config_repo_result";
+import { ConfigRepoVM, CRVMAware, WebhookUrlGenerator } from "./config_repo_view_model";
 import styles from "./index.scss";
+import { WebhookSuggestions } from "./webhook_suggestions";
 
 interface CollectionAttrs extends RequiresPluginInfos {
   models: Stream<ConfigRepoVM[]>;
   flushEtag: () => void;
   sm: ScrollManager;
+  page: WebhookUrlGenerator;
```
[the remark this answers is not in the document]
Comment (marques-work) [reply server/src/main/webapp/WEB-INF/rails/webpack/views/pages/config_repos/config_repos_widget.tsx:44]: True, I suppose I was looking at it from a different perspective. More descriptive.

**Pull request merged**